modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/bookmark/components/UpsellViewModel.kt

```diff
@@ -25,6 +28,8 @@ import javax.inject.Inject
 class UpsellViewModel @Inject constructor(
     private val analyticsTracker: AnalyticsTrackerWrapper,
     private val subscriptionManager: SubscriptionManager,
+    private val feature: FeatureWrapper,
+    private val releaseVersion: ReleaseVersionWrapper,
 ) : ViewModel() {
```

Comment:
Instead of having the `FeatureWrapper` class with a hardcoded reference to the bookmarks feature, we could just have an alternate constructor for testing purposes:

``` kotlin
class UpsellViewModel @VisibleForTesting internal constructor(
    private val analyticsTracker: AnalyticsTrackerWrapper,
    private val subscriptionManager: SubscriptionManager,
    private val bookmarksFeature: Feature,
    private val releaseVersion: ReleaseVersionWrapper,
) : ViewModel() {

    @Inject
    constructor(
        analyticsTracker: AnalyticsTrackerWrapper,
        subscriptionManager: SubscriptionManager,
        releaseVersion: ReleaseVersionWrapper,
    ) : this(
        analyticsTracker = analyticsTracker,
        subscriptionManager = subscriptionManager,
        bookmarksFeature = Feature.BOOKMARKS_ENABLED,
        releaseVersion = releaseVersion,
    )
```

Sharing this more just as food for thought. I tend to prefer avoiding wrapper classes for testing when possible, but having alternate constructors isn't an ideal solution either since they have a lot of repetition.

The other option that comes to mind would be having a custom method in a hilt module that takes in the injected dependencies and uses them to manually construct and return the `UpsellViewModel`. That has the benefit of not bloating the `UpsellViewModel` class, but I'm not sure if adding that bit of DI-indirection might end up making this seem more confusing.

Long story short, if I were writing this I probably would have used one of the alternatives that avoided using a wrapper class, but I realize the alternatives have their own problems, so no need to change this, I'm just talking about loud here.

Reply:
Thanks for sharing your thoughts! This was really tricky to test and each approach seemed to have its own pros and cons. At this time, I'd like to continue with the wrapper classes but I'll be on the lookout for a clearer testing approach that can be addressed in a future PR.